File: autobyteus-android/app/src/main/java/org/autobyteus/mobile/web/TrustedNavigationPolicy.kt
package org.autobyteus.mobile.web

import org.autobyteus.mobile.connection.SavedNodeProfile
import java.net.URI

enum class NavigationDecisionType {
    AllowInWebView,
    OpenExternal,
    Block,
}

data class NavigationDecision(
    val type: NavigationDecisionType,
    val reason: String,
)

object TrustedNavigationPolicy {
    fun classify(targetUrl: String, profile: SavedNodeProfile): NavigationDecision {
        val uri = try {
            URI(targetUrl)
        } catch (error: Exception) {
            return NavigationDecision(NavigationDecisionType.Block, "URL could not be parsed")
        }
        val scheme = uri.scheme?.lowercase()
            ?: return NavigationDecision(NavigationDecisionType.Block, "URL has no scheme")

        if (scheme == "mailto" || scheme == "tel" || scheme == "sms") {
            return NavigationDecision(NavigationDecisionType.OpenExternal, "Non-web link opens outside AutoByteus")
        }
        if (scheme != "http" && scheme != "https") {
            return NavigationDecision(NavigationDecisionType.Block, "Unsupported scheme")
        }

        val host = uri.host?.lowercase()
            ?: return NavigationDecision(NavigationDecisionType.Block, "URL has no host")
        val sameOrigin = scheme == profile.scheme.lowercase() &&
            host == profile.host.lowercase() &&
            normalizedPort(scheme, uri.port) == normalizedPort(profile.scheme, profile.port ?: -1)
        if (!sameOrigin) {
            return NavigationDecision(NavigationDecisionType.OpenExternal, "Different origin opens outside the WebView")
        }

        val path = uri.rawPath?.ifBlank { "/" } ?: "/"
        if (!isAllowedAutoByteusPath(path)) {
            return NavigationDecision(NavigationDecisionType.Block, "Same-origin path is outside the mobile shell allowlist")
        }
        return NavigationDecision(NavigationDecisionType.AllowInWebView, "Trusted AutoByteus mobile origin")
    }

    private fun isAllowedAutoByteusPath(path: String): Boolean {
        val cleanPath = path.ifBlank { "/" }
        return cleanPath == "/mobile" ||
            cleanPath.startsWith("/mobile/") ||
            cleanPath.startsWith("/rest/") ||
            cleanPath == "/graphql" ||
            cleanPath.startsWith("/graphql/") ||
            cleanPath.startsWith("/_nuxt/") ||
            cleanPath.startsWith("/__nuxt/") ||
            cleanPath.startsWith("/assets/") ||
            cleanPath == "/mobile.webmanifest" ||
            cleanPath == "/autobyteus-icon.svg" ||
            cleanPath == "/favicon.ico"
    }

    private fun normalizedPort(scheme: String, port: Int): Int = when {
        port >= 0 -> port
        scheme.equals("https", ignoreCase = true) -> 443
        scheme.equals("http", ignoreCase = true) -> 80
        else -> port
    }
}

File: autobyteus-ios/AutoByteusMobileCore/TrustedNavigationPolicy.swift
import Foundation

public enum NavigationDecisionType: String, Codable, Equatable {
    case allowInWebView
    case openExternal
    case block
}

public struct NavigationDecision: Codable, Equatable {
    public let type: NavigationDecisionType
    public let reason: String

    public init(type: NavigationDecisionType, reason: String) {
        self.type = type
        self.reason = reason
    }
}

public enum TrustedNavigationPolicy {
    public static func classify(targetURL: String, profile: SavedNodeProfile) -> NavigationDecision {
        guard let components = URLComponents(string: targetURL) else {
            return NavigationDecision(type: .block, reason: "URL could not be parsed")
        }
        guard let scheme = components.scheme?.lowercased() else {
            return NavigationDecision(type: .block, reason: "URL has no scheme")
        }

        if ["mailto", "tel", "sms"].contains(scheme) {
            return NavigationDecision(type: .openExternal, reason: "Non-web link opens outside AutoByteus")
        }
        guard scheme == "http" || scheme == "https" else {
            return NavigationDecision(type: .block, reason: "Unsupported scheme")
        }
        guard let host = components.host?.lowercased() else {
            return NavigationDecision(type: .block, reason: "URL has no host")
        }

        let sameOrigin = scheme == profile.scheme.lowercased()
            && host == profile.host.lowercased()
            && normalizedPort(scheme: scheme, port: components.port) == normalizedPort(scheme: profile.scheme, port: profile.port)
        guard sameOrigin else {
            return NavigationDecision(type: .openExternal, reason: "Different origin opens outside the WebView")
        }

        let path = components.percentEncodedPath.isEmpty ? "/" : components.percentEncodedPath
        guard isAllowedAutoByteusPath(path) else {
            return NavigationDecision(type: .block, reason: "Same-origin path is outside the mobile shell allowlist")
        }
        return NavigationDecision(type: .allowInWebView, reason: "Trusted AutoByteus mobile origin")
    }

    private static func isAllowedAutoByteusPath(_ path: String) -> Bool {
        let cleanPath = path.isEmpty ? "/" : path
        return cleanPath == "/mobile" ||
            cleanPath.hasPrefix("/mobile/") ||
            cleanPath.hasPrefix("/rest/") ||
            cleanPath == "/graphql" ||
            cleanPath.hasPrefix("/graphql/") ||
            cleanPath.hasPrefix("/_nuxt/") ||
            cleanPath.hasPrefix("/__nuxt/") ||
            cleanPath.hasPrefix("/assets/") ||
            cleanPath == "/mobile.webmanifest" ||
            cleanPath == "/autobyteus-icon.svg" ||
            cleanPath == "/favicon.ico"
    }

    private static func normalizedPort(scheme: String, port: Int?) -> Int {
        if let port { return port }
        if scheme.caseInsensitiveCompare("https") == .orderedSame { return 443 }
        if scheme.caseInsensitiveCompare("http") == .orderedSame { return 80 }
        return -1
    }
}
